File: app/recommender.py
"""
Core recommendation engine: ingredient-based recipe retrieval using
TF-IDF vectorization + cosine similarity over the recipe knowledge base,
with hard filters for diet, cuisine and cooking time.
"""
import json
import os
import re

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

DATA_PATH = os.path.join(os.path.dirname(__file__), "..", "data", "recipes.json")

# A small synonym map so casual user phrasing ("tomatoes", "paneer cheese")
# lines up with the normalized ingredient vocabulary used in the dataset.
SYNONYMS = {
    "tomatoes": "tomato",
    "onions": "onion",
    "potatoes": "potato",
    "eggplants": "eggplant",
    "aubergine": "eggplant",
    "brinjal": "eggplant",
    "capsicum": "bell pepper",
    "bell peppers": "bell pepper",
    "peppers": "bell pepper",
    "cottage cheese": "paneer",
    "garbanzo beans": "chickpeas",
    "garbanzos": "chickpeas",
    "chick peas": "chickpeas",
    "spring onions": "spring onion",
    "scallion": "spring onion",
    "scallions": "spring onion",
    "coriander": "coriander powder",
    "egg": "eggs",
    "prawns": "shrimp",
    "prawn": "shrimp",
    "beef mince": "ground beef",
    "minced beef": "ground beef",
    "mince": "ground beef",
    "curd": "yogurt",
    "cheese": "cheese",
}


def _normalize_token(text):
    text = text.strip().lower()
    text = re.sub(r"[^a-z\s]", "", text)
    text = re.sub(r"\s+", " ", text).strip()
    return SYNONYMS.get(text, text)


def _to_doc(ingredients):
    """Turn a list of ingredient strings into an underscore-joined
    token string, e.g. ['red lentils', 'onion'] -> 'red_lentils onion'."""
    return " ".join(ing.strip().lower().replace(" ", "_") for ing in ingredients)


class RecipeRecommender:
    def __init__(self, data_path=DATA_PATH):
        with open(data_path) as f:
            self.recipes = json.load(f)

        self.corpus = [_to_doc(r["ingredients"]) for r in self.recipes]
        # Each ingredient (possibly multi-word, underscored) is one token,
        # so the default whitespace tokenizer is exactly what we want.
        self.vectorizer = TfidfVectorizer(token_pattern=r"[^\s]+")
        self.matrix = self.vectorizer.fit_transform(self.corpus)

    def normalize_ingredients(self, raw_ingredients):
        return [_normalize_token(i) for i in raw_ingredients if _normalize_token(i)]

    def recommend(self, user_ingredients, diet=None, cuisine=None,
                   max_minutes=None, top_k=5):
        """
        user_ingredients: list of raw ingredient strings the user has.
        diet: optional string, e.g. "vegetarian", "vegan", "gluten-free".
        cuisine: optional string, e.g. "Indian", "Italian".
        max_minutes: optional int, max ready_in_minutes.
        Returns a list of result dicts sorted by relevance.
        """
        norm_ingredients = self.normalize_ingredients(user_ingredients)
        query_doc = _to_doc(norm_ingredients) if norm_ingredients else ""
        query_vec = self.vectorizer.transform([query_doc])
        sims = cosine_similarity(query_vec, self.matrix).flatten()

        user_set = set(norm_ingredients)
        results = []
        for idx, recipe in enumerate(self.recipes):
            if diet and diet.lower() not in [d.lower() for d in recipe["diet"]]:
                continue
            if cuisine and cuisine.lower() != recipe["cuisine"].lower():
                continue
            if max_minutes and recipe["ready_in_minutes"] > max_minutes:
                continue

            recipe_set = set(recipe["ingredients"])
            matched = sorted(user_set & recipe_set)
            missing = sorted(recipe_set - user_set)
            coverage = len(matched) / len(recipe_set) if recipe_set else 0.0

            results.append({
                "recipe": recipe,
                "similarity": float(sims[idx]),
                "matched_ingredients": matched,
                "missing_ingredients": missing,
                "coverage": coverage,
            })

        # Rank primarily by how many of the recipe's ingredients the user
        # already has (coverage), then by TF-IDF similarity as a tiebreaker
        # so distinctive shared ingredients (e.g. "paneer") count for more
        # than common ones (e.g. "onion").
        results.sort(key=lambda r: (r["coverage"], r["similarity"]), reverse=True)
        return results[:top_k]

    def available_cuisines(self):
        return sorted({r["cuisine"] for r in self.recipes})

    def available_diets(self):
        diets = set()
        for r in self.recipes:
            diets.update(r["diet"])
        return sorted(diets)

File: lib/recommender.js
/**
 * Core recommendation engine: ingredient-based recipe retrieval.
 *
 * How it scores a recipe against what the user has:
 *  - "coverage": what fraction of the RECIPE's ingredients the user already
 *    has (e.g. user has 4 of the 9 ingredients -> 0.44). This is the main
 *    ranking signal, since it directly answers "what can I actually cook?".
 *  - "similarity": a weighted overlap score where rare ingredients across
 *    the whole dataset (e.g. "paneer", "shrimp") count for more than common
 *    ones (e.g. "onion", "garlic"). This is used as a tiebreaker so two
 *    recipes with the same coverage rank sensibly. It's the same idea as
 *    TF-IDF weighting, done directly on a small ingredient vocabulary
 *    instead of full free-text.
 */
import recipes from "./recipes.json" with { type: "json" };

const SYNONYMS = {
  tomatoes: "tomato",
  onions: "onion",
  potatoes: "potato",
  eggplants: "eggplant",
  aubergine: "eggplant",
  brinjal: "eggplant",
  capsicum: "bell pepper",
  "bell peppers": "bell pepper",
  peppers: "bell pepper",
  "cottage cheese": "paneer",
  "garbanzo beans": "chickpeas",
  garbanzos: "chickpeas",
  "chick peas": "chickpeas",
  "spring onions": "spring onion",
  scallion: "spring onion",
  scallions: "spring onion",
  coriander: "coriander powder",
  egg: "eggs",
  prawns: "shrimp",
  prawn: "shrimp",
  "beef mince": "ground beef",
  "minced beef": "ground beef",
  mince: "ground beef",
  curd: "yogurt",
};

function normalizeToken(raw) {
  const text = raw
    .trim()
    .toLowerCase()
    .replace(/[^a-z\s]/g, "")
    .replace(/\s+/g, " ")
    .trim();
  return SYNONYMS[text] || text;
}

// Document-frequency weight for each ingredient across the dataset: an
// ingredient that appears in few recipes (paneer) gets a higher weight than
// one that appears in most of them (onion, garlic).
function buildIngredientWeights() {
  const docCount = {};
  for (const recipe of recipes) {
    for (const ing of new Set(recipe.ingredients)) {
      docCount[ing] = (docCount[ing] || 0) + 1;
    }
  }
  const weights = {};
  for (const [ing, count] of Object.entries(docCount)) {
    weights[ing] = Math.log(recipes.length / count + 1);
  }
  return weights;
}

const INGREDIENT_WEIGHTS = buildIngredientWeights();

function weightedSimilarity(userSet, recipeIngredients) {
  let score = 0;
  let recipeNorm = 0;
  for (const ing of recipeIngredients) {
    const w = INGREDIENT_WEIGHTS[ing] || 1;
    recipeNorm += w * w;
    if (userSet.has(ing)) score += w * w;
  }
  if (recipeNorm === 0) return 0;
  return score / Math.sqrt(recipeNorm);
}

export function normalizeIngredients(rawIngredients) {
  return rawIngredients.map(normalizeToken).filter(Boolean);
}

/**
 * @param {string[]} userIngredients - raw ingredient strings the user typed
 * @param {{ diet?: string, cuisine?: string, maxMinutes?: number, topK?: number }} opts
 */
export function recommend(userIngredients, opts = {}) {
  const { diet, cuisine, maxMinutes, topK = 5 } = opts;
  const normalized = normalizeIngredients(userIngredients);
  const userSet = new Set(normalized);

  const results = [];
  for (const recipe of recipes) {
    if (diet && !recipe.diet.some((d) => d.toLowerCase() === diet.toLowerCase())) {
      continue;
    }
    if (cuisine && recipe.cuisine.toLowerCase() !== cuisine.toLowerCase()) {
      continue;
    }
    if (maxMinutes && recipe.ready_in_minutes > maxMinutes) {
      continue;
    }

    const recipeSet = new Set(recipe.ingredients);
    const matched = [...userSet].filter((i) => recipeSet.has(i)).sort();
    const missing = [...recipeSet].filter((i) => !userSet.has(i)).sort();
    const coverage = recipeSet.size ? matched.length / recipeSet.size : 0;
    const similarity = weightedSimilarity(userSet, recipe.ingredients);

    results.push({
      recipe,
      coverage,
      similarity,
      matchedIngredients: matched,
      missingIngredients: missing,
    });
  }

  results.sort((a, b) => b.coverage - a.coverage || b.similarity - a.similarity);
  return results.slice(0, topK);
}

export function availableCuisines() {
  return [...new Set(recipes.map((r) => r.cuisine))].sort();
}

export function availableDiets() {
  const diets = new Set();
  for (const r of recipes) for (const d of r.diet) diets.add(d);
  return [...diets].sort();
}
